### PR TITLE
[REVIEW] update sssp and bfs with new graph class

### PR DESCRIPTION
**Pull request by xcadet**
Remove the Graph and gdf_column from sssp and bfs methods. Use the new GraphCSR object instead.

### CODE REVIEW

**Comment by xcadet** [comment]
rerun tests

**Comment by xcadet** [comment]
rerun tests

File: python/cugraph/traversal/sssp_wrapper.pyx [hunk 1]
```diff
@@ -32,77 +33,112 @@ import cudf._lib as libcudf
...
+    # Step 9: Renumber before return
+    #         It is only required to renumber vertex and predecessors
     if input_graph.renumbered:
-        if isinstance(input_graph.edgelist.renumber_map, cudf.DataFrame):
-            n_cols = len(input_graph.edgelist.renumber_map.columns) - 1
-            unrenumbered_df_ = df.merge(input_graph.edgelist.renumber_map, left_on='vertex', right_on='id', how='left').drop(['id', 'vertex'])
-            unrenumbered_df = unrenumbered_df_.merge(input_graph.edgelist.renumber_map, left_on='predecessor', right_on='id', how='left').drop(['id', 'predecessor'])
-            unrenumbered_df.columns = ['distance']+['vertex_'+str(i) for i in range(n_cols)]+['predecessor_'+str(i) for i in range(n_cols)]
-            cols = unrenumbered_df.columns.to_list()
-            df = unrenumbered_df[cols[1:n_cols+1] + [cols[0]] + cols[n_cols:]]
```
Comment (xcadet) [inline python/cugraph/traversal/sssp_wrapper.pyx]: Is this still supposed to be handled?

File: python/cugraph/traversal/bfs_wrapper.pyx [hunk 2]
```diff
@@ -17,72 +17,88 @@
...
+    # Step 9: Renumber before return
+    #         It is only required to renumber vertex and predecessors
     if input_graph.renumbered:
-        if isinstance(input_graph.edgelist.renumber_map, cudf.DataFrame):
-            n_cols = len(input_graph.edgelist.renumber_map.columns) - 1
-            unrenumbered_df_ = df.merge(input_graph.edgelist.renumber_map, left_on='vertex', right_on='id', how='left').drop(['id', 'vertex'])
-            unrenumbered_df = unrenumbered_df_.merge(input_graph.edgelist.renumber_map, left_on='predecessor', right_on='id', how='left').drop(['id', 'predecessor'])
-            unrenumbered_df.columns = ['distance']+['vertex_'+str(i) for i in range(n_cols)]+['predecessor_'+str(i) for i in range(n_cols)]
-            cols = unrenumbered_df.columns.to_list()
-            df = unrenumbered_df[cols[1:n_cols+1] + [cols[0]] + cols[n_cols:]]
```
Comment (xcadet) [inline python/cugraph/traversal/bfs_wrapper.pyx]: Is this still supposed to handled?

File: cpp/include/algorithms.hpp [hunk 3]
```diff
@@ -257,4 +257,57 @@ void k_truss_subgraph(experimental::GraphCOO<VT, ET, WT> const &graph,
...
+ * @param[in] start_vertex           The starting vertex for SSSP
+ *
+ */
+/* ----------------------------------------------------------------------------*/
+template <typename VT, typename ET, typename WT>
+void sssp(experimental::GraphCSR<VT,ET,WT> const &graph,
+          WT *distances,
+          VT *predecessors,
+          const VT source_vertex);
+/**
```
Comment (BradReesWork) [inline cpp/include/algorithms.hpp]: please put a blank line between function definitions to improve readability

File: cpp/src/traversal/bfs.cu [hunk 4]
```diff
@@ -469,29 +471,26 @@ namespace detail {
...
-  CUGRAPH_EXPECTS(graph->adjList != nullptr, "Invalid API parameter");
-  CUGRAPH_EXPECTS(graph->adjList->offsets->dtype == GDF_INT32, "Unsupported data type");
-  CUGRAPH_EXPECTS(graph->adjList->indices->dtype == GDF_INT32, "Unsupported data type");
+} // !namespace cugraph::detail
+
+template <typename VT, typename ET, typename WT>
+void bfs(experimental::GraphCSR<VT, ET, WT> const &graph, VT *distances, VT *predecessors, const VT start_vertex, bool directed) {
   CUGRAPH_EXPECTS(typeid(VT) == typeid(int), "Unsupported data type");
+  CUGRAPH_EXPECTS(typeid(ET) == typeid(int), "Unsupported data type");
+  CUGRAPH_EXPECTS((typeid(WT) == typeid(float)) || (typeid(WT) == typeid(double)), "Unsupported data type");
```
Comment (BradReesWork) [inline cpp/src/traversal/bfs.cu]: please expand error messages - there is no indication which argument or variable is of an unsupported type

File: python/cugraph/traversal/sssp_wrapper.pyx (hunk 1 above)
Comment (afender) [reply python/cugraph/traversal/sssp_wrapper.pyx]: yes, that looks like the unrenumbering step

File: python/cugraph/traversal/bfs_wrapper.pyx (hunk 2 above)
Comment (afender) [reply python/cugraph/traversal/bfs_wrapper.pyx]: yes, that looks like the unrenumbering step

File: cpp/src/traversal/sssp.cu [hunk 5]
```diff
@@ -314,91 +310,42 @@ void sssp(Graph* graph,
...
-                          cudaMemcpyHostToDevice));
-      gdf_column_view(graph->adjList->edge_data,
-                      d_edge_data,
-                      nullptr,
-                      graph->adjList->indices->size,
-                      GDF_FLOAT32);
-    }
+    std::vector<WT> h_edge_data(graph.number_of_edges, static_cast<WT>(1));
+    size_t edge_data_size = sizeof(WT) * h_edge_data.size();
+    ALLOC_TRY((WT**)&edge_weights_ptr, edge_data_size, stream);
```
Comment (kaatish) [inline cpp/src/traversal/sssp.cu]: Try to use rmm::device_vector or device_buffer here

File: cpp/src/traversal/sssp.cu [hunk 6]
```diff
@@ -314,91 +310,42 @@ void sssp(Graph* graph,
...
-    // TODO fix me after gdf_column is removed from Graph
-    // if (distances) CUGRAPH_EXPECTS(typeid(distances) == typeid(graph->adjList->edge_data), "distances and weights type mismatch");
-
     // SSSP is not defined for graphs with negative weight cycles
     // Warn user about any negative edges
-    if (graph->prop && graph->prop->has_negative_edges == GDF_PROP_TRUE)
+    if (graph.prop.has_negative_edges == experimental::PropType::PROP_TRUE)
       std::cerr << "WARN: The graph has negative weight edges. SSSP will not "
                    "converge if the graph has negative weight cycles\n";
+    edge_weights_ptr = (WT*)graph.edge_data;
```
Comment (afender) [inline cpp/src/traversal/sssp.cu]: Why do we cast to `WT*` here?
We are pretty sure `graph.edge_data` is `WT*` at this point otherwise it would be detected at compile time it seems.

File: cpp/src/traversal/sssp.cu [hunk 7]
```diff
@@ -314,91 +310,42 @@ void sssp(Graph* graph,
...
 
-  int n = graph->adjList->offsets->size - 1;
-  int e = graph->adjList->indices->size;
-  int* offsets_ptr = (int*)graph->adjList->offsets->data;
-  int* indices_ptr = (int*)graph->adjList->indices->data;
+  int n = graph.number_of_vertices;
+  int e = graph.number_of_edges;
 
-  void* edge_weights_ptr = static_cast<void*>(graph->adjList->edge_data->data);
+  VT* offsets_ptr = (VT*)graph.offsets;
```
Comment (afender) [inline cpp/src/traversal/sssp.cu]:
```suggestion
  ET* offsets_ptr = graph.offsets;
```
Should be `ET*`?
Is cast needed?

File: cpp/tests/sssp/sssp_test.cu [hunk 8]
```diff
@@ -397,34 +391,48 @@ class Tests_SSSP : public ::testing::TestWithParam<SSSP_Usecase> {
...
-    ALLOC_FREE_TRY(col_dest.data, stream);
-    ALLOC_FREE_TRY(col_weights.data, stream);
   }
 };
 
 std::vector<double> Tests_SSSP::SSSP_time;
 
-TEST_P(Tests_SSSP, CheckFP32_DIST_NO_PREDS) {
-  run_current_test<int, int, float, true, false>(GetParam());
+TEST_P(Tests_SSSP, CheckFP32_NO_RANDOM_DIST_NO_PREDS) {
```
Comment (afender) [inline cpp/tests/sssp/sssp_test.cu]: Can we add a comment on what is "random" or not in these tests?

File: python/cugraph/tests/test_bfs.py [hunk 9]
```diff
@@ -97,13 +97,10 @@ def test_bfs(managed, pool, graph_file):
...
-    # assert len(base_dist) == len(cugraph_dist)
     i = 0
     j = 0
     while i < len(cugraph_dist):
-        if base_vid[i] == cugraph_vid[i]:
-            assert base_dist[i] == cugraph_dist[i]
-        else:
-            j = j+1
-        i = i+1
+        if base_vid[j] == cugraph_vid[i]:
```
Comment (afender) [inline python/cugraph/tests/test_bfs.py]: Not sure about the logic here. The previous code was supposed to count the number of mismatches in `j`. What does this code do?

Also, there is room for improving the test here as none actually checks the number of mismatches.

File: python/cugraph/traversal/bfs_wrapper.pyx [hunk 10]
```diff
@@ -17,72 +17,88 @@
...
+    # Step 3: Extract CSR offsets, indices, weights are not expected
+    #         - offsets: int (signed, 32-bit)
+    #         - indices: int (signed, 32-bit)
+    [offsets, indices] = graph_wrapper.datatype_cast([input_graph.adjlist.offsets, input_graph.adjlist.indices], [np.int32])
+    c_offsets_ptr = offsets.__cuda_array_interface__['data'][0]
+    c_indices_ptr = indices.__cuda_array_interface__['data'][0]
 
+    # Step 4: Setup number of vertices and edges
+    num_verts = input_graph.number_of_vertices()
+    num_edges = len(indices)
```
Comment (afender) [inline python/cugraph/traversal/bfs_wrapper.pyx]: input_graph.number_of_vertices()
```suggestion
    num_edges = input_graph.number_of_edges()
```

File: python/cugraph/traversal/sssp_wrapper.pyx [hunk 11]
```diff
@@ -32,77 +33,112 @@ import cudf._lib as libcudf
...
+    c_indices_ptr = indices.__cuda_array_interface__['data'][0]
+
+    data_type = np.int32
+    if weights is not None:
+        data_type = weights.dtype
+        c_weights_ptr = weights.__cuda_array_interface__['data'][0]
+
+    # Step 4: Setup number of vertices and number of edges
+    num_verts = input_graph.number_of_vertices()
+    num_edges = len(indices)
```
Comment (afender) [inline python/cugraph/traversal/sssp_wrapper.pyx]: here as well

File: cpp/src/traversal/sssp.cu (hunk 6 above)
Comment (afender) [reply cpp/src/traversal/sssp.cu]: If it is about the const, this may go away in https://github.com/rapidsai/cugraph/pull/780 (also being discussed there)

File: python/cugraph/tests/test_bfs.py (hunk 9 above)
Comment (Iroy30) [reply python/cugraph/tests/test_bfs.py]: Xavier and I had a conversation on the changes in the test. The test asserts on every comparison of distance. If we want to track the number of mismatches, the assertion on 104 could be changed to an err counter

File: python/cugraph/traversal/bfs_wrapper.pyx (hunk 10 above)
Comment (Iroy30) [reply python/cugraph/traversal/bfs_wrapper.pyx]: This change could lead to errors for some cases. In case the user created a Graph() type with edgelist, input_graph.number_of_edges() gives the number of undirected edges now.

File: python/cugraph/traversal/bfs_wrapper.pyx (hunk 2 above)
Comment (Iroy30) [reply python/cugraph/traversal/bfs_wrapper.pyx]: The lines 77-88 are for multicolumn and single column unrenumbering. Better to leave it as is. It is subject to changes in further releases as we change our renumbering/unrenumbering strategy.

File: python/cugraph/traversal/bfs_wrapper.pyx [hunk 12]
```diff
@@ -17,72 +17,88 @@
...
-            unrenumbered_df.columns = ['distance']+['vertex_'+str(i) for i in range(n_cols)]+['predecessor_'+str(i) for i in range(n_cols)]
-            cols = unrenumbered_df.columns.to_list()
-            df = unrenumbered_df[cols[1:n_cols+1] + [cols[0]] + cols[n_cols:]]
-        else:
-            df['vertex'] = input_graph.edgelist.renumber_map[df['vertex']]
-            df['predecessor'][df['predecessor']>-1] = input_graph.edgelist.renumber_map[df['predecessor'][df['predecessor']>-1]]
-
-    return df
+        df = unrenumber(input_graph.edgelist.renumber_map, df, 'vertex')
+        df['predecessor'][df['predecessor'] > -1] = input_graph.edgelist.renumber_map[df['predecessor'][df['predecessor'] > -1]]
```
Comment (Iroy30) [inline python/cugraph/traversal/bfs_wrapper.pyx]: This line would give an error when there is multicolumn unrenumbering when renumber_map is a dataframe

File: python/cugraph/traversal/sssp_wrapper.pyx (hunk 1 above)
Comment (xcadet) [reply python/cugraph/traversal/sssp_wrapper.pyx]: Reverted it in https://github.com/rapidsai/cugraph/pull/764/commits/1528aebe9ea39d6ac0060b1068a76f8c2a36056c

File: python/cugraph/traversal/bfs_wrapper.pyx (hunk 2 above)
Comment (xcadet) [reply python/cugraph/traversal/bfs_wrapper.pyx]: Reverted it in https://github.com/rapidsai/cugraph/pull/764/commits/1528aebe9ea39d6ac0060b1068a76f8c2a36056c, and added a FIXME about it

File: cpp/include/algorithms.hpp (hunk 3 above)
Comment (xcadet) [reply cpp/include/algorithms.hpp]: Added in https://github.com/rapidsai/cugraph/pull/764/commits/1528aebe9ea39d6ac0060b1068a76f8c2a36056c

File: cpp/src/traversal/bfs.cu (hunk 4 above)
Comment (xcadet) [reply cpp/src/traversal/bfs.cu]: Added in https://github.com/rapidsai/cugraph/pull/764/commits/1528aebe9ea39d6ac0060b1068a76f8c2a36056c, now includes indications and which types can be used

File: cpp/src/traversal/sssp.cu (hunk 5 above)
[the remark this answers is not in the document]
Comment (xcadet) [reply cpp/src/traversal/sssp.cu]: Updated in https://github.com/rapidsai/cugraph/pull/764/commits/1528aebe9ea39d6ac0060b1068a76f8c2a36056c, now uses `thrust::device_vector`

File: cpp/src/traversal/sssp.cu (hunk 6 above)
Comment (xcadet) [reply cpp/src/traversal/sssp.cu]: Udpated in https://github.com/rapidsai/cugraph/pull/764/commits/1528aebe9ea39d6ac0060b1068a76f8c2a36056c: SSSP and BFS now use const pointers

File: cpp/src/traversal/sssp.cu (hunk 7 above)
Comment (xcadet) [reply cpp/src/traversal/sssp.cu]: Updated in https://github.com/rapidsai/cugraph/pull/764/commits/1528aebe9ea39d6ac0060b1068a76f8c2a36056c, changed `VT*` to `const ET*`

File: cpp/tests/sssp/sssp_test.cu (hunk 8 above)
Comment (xcadet) [reply cpp/tests/sssp/sssp_test.cu]: Added comments above definition of `run_current_test` in https://github.com/rapidsai/cugraph/pull/764/commits/1528aebe9ea39d6ac0060b1068a76f8c2a36056c

File: python/cugraph/tests/test_bfs.py (hunk 9 above)
Comment (xcadet) [reply python/cugraph/tests/test_bfs.py]: Renamed variables and moved assert to the end with a counter in https://github.com/rapidsai/cugraph/pull/764/commits/1528aebe9ea39d6ac0060b1068a76f8c2a36056c

File: python/cugraph/traversal/bfs_wrapper.pyx (hunk 12 above)
Comment (xcadet) [reply python/cugraph/traversal/bfs_wrapper.pyx]: Reverted to include verification of the type of `input_grapg.edgelist.renumber_map` and added a FIXME in https://github.com/rapidsai/cugraph/pull/764/commits/1528aebe9ea39d6ac0060b1068a76f8c2a36056c

File: cpp/src/traversal/sssp.cu (hunk 5 above)
Comment (xcadet) [reply cpp/src/traversal/sssp.cu]: Updated in https://github.com/rapidsai/cugraph/commit/1528aebe9ea39d6ac0060b1068a76f8c2a36056c, uses thrust::device_vector instead

File: cpp/src/traversal/sssp.cu [hunk 13]
```diff
@@ -314,91 +310,42 @@ void sssp(Graph* graph,
...
-                      GDF_FLOAT32);
-    }
+    std::vector<WT> h_edge_data(graph.number_of_edges, static_cast<WT>(1));
+    size_t edge_data_size = sizeof(WT) * h_edge_data.size();
+    ALLOC_TRY((WT**)&edge_weights_ptr, edge_data_size, stream);
+    CUDA_TRY(cudaMemcpy(edge_weights_ptr,
+                        &h_edge_data[0],
+                        edge_data_size,
+                        cudaMemcpyHostToDevice));
+    should_free_weights = true;
```
[the remark this answers is not in the document]
Comment (xcadet) [reply cpp/src/traversal/sssp.cu]: Updated in https://github.com/rapidsai/cugraph/commit/1528aebe9ea39d6ac0060b1068a76f8c2a36056c, `should_free_weights` is no longer used

File: python/cugraph/traversal/bfs_wrapper.pyx (hunk 10 above)
Comment (afender) [reply python/cugraph/traversal/bfs_wrapper.pyx]: Got it.

File: python/cugraph/traversal/sssp_wrapper.pyx (hunk 11 above)
Comment (afender) [reply python/cugraph/traversal/sssp_wrapper.pyx]: please disregard
